source/beamline/CLS/CLSBeamline.cpp

```diff
@@ -1,12 +1,227 @@
+		emit connected(wasConnected_);
+	}
 }
 
-CLSBeamline::~CLSBeamline()
+void CLSBeamline::updateBeamStatus()
 {
+	bool beamOn = CLSStorageRing::sr1()->beamAvailable();
+	if (beamlineStatus_) {
+		beamOn &= beamlineStatus_->isOn();
+	}
+
+	if (wasBeamOn_ != beamOn) {
+		wasBeamOn_ = beamOn;
+		emit beamAvaliabilityChanged(beamOn);
+	}
+}
+
+void CLSBeamline::onBeamOnActionFinished(){
+	disconnect( beamOnAction_, 0, this, 0 );
+
+	beamOnAction_->deleteLater();
+	beamOnAction_ = 0; //NULL
+}
+
+void CLSBeamline::onBeamOnActionFailed(){
+	AMErrorMon::error(this, CLSBEAMLINE_ERR_BEAM_ON_FAILED, QString("Failed to execute the beam on actions with message: %1.").arg(beamOnAction_->failureMessage()), true);
+	onBeamOnActionFinished();
+}
+
+void CLSBeamline::onBeamOffActionFinished(){
+	disconnect( beamOffAction_, 0, this, 0 );
+
+	beamOffAction_->deleteLater();
+	beamOffAction_ = 0; //NULL
+}
+
+/// ==================== protected methods =======================
+
+void CLSBeamline::createBeamlineStatus(CLSShutters *shutters, CLSValves *valves)
+{
+	if (shutters)
+		beamlineShutters_ = shutters;
+	else
+		beamlineShutters_ = new CLSShutters(QString("%1 Shutters").arg(beamlineName_), this);
+
+	if (valves)
+		beamlineValves_ = valves;
+	else
+		beamlineValves_ = new CLSValves(QString("%1 Valves").arg(beamlineName_), this);
+
+	CLSBeamlineStatus *beamlineStatus = new CLSBeamlineStatus(QString("%1 BeamlineStatus").arg(beamlineName_), this);
+	beamlineStatus->addShutterControl(beamlineShutters_, CLSShutters::Open);
+	beamlineStatus->addValveControl(beamlineValves_, CLSValves::Open);
+
+	setBeamlineStatus(beamlineStatus);
+}
+
+void CLSBeamline::initializeBeamline()
+{
+	if (beamlineStatus_ && beamlineStatus_->isConnected()) {
+		updateBeamStatus();
+	}
+
+	onBeamlineComponentConnected();
+}
+
+void CLSBeamline::setBeamlineStatus(CLSBeamlineStatus *beamlineStatus)
+{
+	if (beamlineStatus_ != beamlineStatus) {
+		if (beamlineStatus_) {
+			requiredControls_->removeControl(beamlineStatus_);
+			disconnect(beamlineStatus, 0, this, 0);
+		}
+
+		beamlineStatus_ = beamlineStatus;
+		if (beamlineStatus_) {
+			requiredControls_->addControl(beamlineStatus_);
+
+			connect(beamlineStatus_, SIGNAL(beamStatusChanged(bool)), this, SLOT(updateBeamStatus()));
+			connect(beamlineStatus_, SIGNAL(connected(bool)), this, SLOT(updateBeamStatus()));
+		}
+
+		updateBeamStatus();
+	}
+}
+
+AMListAction3* CLSBeamline::createBeamOnActions() const
+{
+	if (beamlineStatus_->isOn()) {
+		AMErrorMon::error(this, CLSBEAMLINE_ERR_BEAM_ON_ALREADY_ON, QString("Failed to create the beam on actions because the beam is already ready."), true);
+		return 0;
+	}
+
+	if(!beamlineShutters_->isConnected() || !beamlineValves_->isConnected()) {
+		AMErrorMon::error(this, CLSBEAMLINE_ERR_BEAM_ON_UNCONNECTED_PV, QString("Failed to create the beam on actions due to unconnected shutter/valve PVs."), true);
+		return 0;
+	}
+
+	if (!beamlineShutters_->isSafetyShutterOpen()) {
```

Comment:
Using VESPERS as an example, beam on would include opening a safety shutter and would therefore never make it past this statement.  Is there a means of bypassing that?